### DISCUSSION TITLE
Alternative to prepared functions

### DISCUSSION THREAD

**Discussion by jakerieger** [planetscale/discussion#242] [Q&A] [answered]
I see that PlanetScale doesn't support prepared functions or procedures. We need to store some metadata for one of our tables that would require the use of a function to calculate the value for each row.

For example, we have a `Posts` table and a `Likes` table that stores user's likes on posts. I want to store the like count on each row in the `Posts` table instead of using a separate query to grab that value in our backend. Normally, you could do this with a function but since it's not supported, is there a workaround for this kind of problem that can be implemented on the DB side?

Main reason is performance. Doing all of these queries separately when grabbing a user's feed takes multiple seconds which is unacceptably slow.

**Comment by orware** [comment, marked as answer]
Hi @jakerieger,

You are correct that we do not provide support for prepared functions/procedures at this time, but hopefully one of the solutions described below may be helpful for you needs!

---

An alternative here if you have the option to do so, would be to create a scheduled task that could periodically run on its own schedule to update the like count value you would like to store in the `Posts` table.

Essentially, rather than running those separate queries for each row on page load for example, via the scheduled task you'd be able to run those separately and precompute things and store the results within your `Posts` table, allowing you to keep all of your needed data within the single query to `Posts`.

The main drawback with this approach is having slightly out of date like counts shown on the page, depending on how frequently that scheduled task is able to run, but in many cases this is an acceptable tradeoff.

---

Another approach I've found helpful to use in situations like this is to think of the data that's about to be presented on the page as an entire set, rather than individual rows, and is something that can work particularly well if you are always dealing with a fairly limited number of results due to using pagination.

For example, let's say you are outputting 10 `Posts` on the page and each of those has their own `Likes`. 

To my understanding, this would lead to the 1 initial query hopefully, that retrieved all 10 `Posts` at once, but now you have to run 10 queries against your `Likes` table to get the like count for each of your `Posts`, leading to 11 queries in total.

If you have the ability to do so, you could create a query, that takes in the 10 Post IDs that are available from the 1 initial query's results, and use the Post IDs to create a query against your `Likes` table that returns the Post IDs + like counts in a single result, so you only have to run 2 queries in total.

Example Query:

```sql
select post_id, count(id) like_count
from likes 
where post_id IN (
    42,
    153,
    904,
    ...
)
GROUP BY post_id
```

This approach wouldn't have the same drawback of the like counts being slightly outdated, since those are still calculated on the fly, but you help avoid the need to run all of those separate queries, especially since I think most ORMs will end up running them all and may not use this approach I just described.

The drawback here is maybe having a slightly more complicated bit of code on the application side to handle this, but overall it shouldn't make things much more complex.

---

I don't know if either of these options may be workable for you, but if not hopefully they may provide some additional ideas!

**Reply by jakerieger** [reply]
I like that second option as we are using pagination for the queries to return the feed. The problem you describe in the first solution is the issue I'm running in to where it only requires one query to grab the posts, but N queries to grab the likes for each post. I also have to grab the comments count as well so that would stack another N queries on top. 

I'm also curious how viable caching the metadata in a Redis DB would be and updating that periodically. Does PlanetScale support triggers? I can use the triggers to update the cache DB as necessary instead of running a service to update on a static timeframe.

**Reply by orware** [reply]
We do not support the use of `TRIGGER` at this time unfortunately. 

To help allow the database to focus more on the data storage / scalability aspects, support for things like procedures/functions/triggers at the database level have been dropped primarily I believe because it has been found to be easier to offload those to the application layer and scale those out horizontally as needed.

Perhaps down the road it might be possible through some new innovations, but at the moment the use of `TRIGGER` definitely falls under the limitations we share over on:
https://planetscale.com/docs/reference/mysql-compatibility

Caching can always be potentially helpful so that could definitely be another approach, and even without the ability to use `TRIGGER`. For example, whenever an update you are going to add a row into your `Likes` table, you can have your application also run the query to calculate the total afterward and update the value back in the `Posts` table for the associated post, which should serve to keep that value always updated and keeps the logic within the application layer as well.

---

Going back to your initial message, I wanted to ask about your last sentence where you mentioned "Doing all of these queries separately when grabbing a user's feed takes multiple seconds which is unacceptably slow".

If there are any extra details regarding whether this type of performance is happening even after a table has been optimized as best as can be expected (indexes, etc.) and can provide a sense of how many rows are being processed (e.g. the number of rows being shared on the screen, along with the associated likes/comments or other information that would also need to be retrieved) that would help provide a little extra insight for any other suggestions I might be able to think of.

**Reply by jakerieger** [reply]
So, we basically need to be able to query from and connect our `Likes`, `Posts`, `Followers`, and `Users` tables when grabbing the feed. The feed is grabbed from the `Posts` table based on posts authored by users the requester follows, so I need to query the `Followers` table first to get all the user IDs. I also need to be able to display the like and comment counts on the feed screen as well. So I'm querying from multiple tables per request. I can join some of the tables in one query but without primary keys this gets a bit harder to do efficiently (at least to my knowledge).

All in all, this API endpoint that grabs the feed currently takes ~6s to return the results from a table that only has about 50 posts currently. Removing the queries to grab the post metadata takes the response time down to ~800ms which is still far too slow but a significant decrease, hence my main concern regarding an efficient way to pull the metadata. Curious what your thoughts are.

**Reply by jakerieger** [reply]
So I decided to build a daemon to update our Redis DB with the metadata we need to store and then reduced the SQL queries down to just one. Overall, response time is down to ~70ms which is much more acceptable. I still think your second suggestion is probably the best way to go down the road but for now, I'm happy with this.

**Reply by orware** [reply]
Nice work @jakerieger! Those results look pretty solid to me with the daemon + Redis approach and it's possible it is a better solution than what my second suggestion would have been able to provide so this looks like a win in my book!
